File: android_mobile/src/main/java/com/alexstyl/specialdates/search/SearchNavigator.java
package com.alexstyl.specialdates.search;

import android.app.Activity;
import android.app.ActivityOptions;
import android.content.Intent;
import android.widget.ImageView;

import com.alexstyl.specialdates.R;
import com.alexstyl.specialdates.analytics.Analytics;
import com.alexstyl.specialdates.contact.Contact;
import com.alexstyl.specialdates.person.PersonActivity;

class SearchNavigator {
    private final Activity activity;
    private final Analytics analytics;

    SearchNavigator(Activity activity, Analytics analytics) {
        this.activity = activity;
        this.analytics = analytics;
    }

    void toContactDetails(Contact contact, ImageView avatar) {
        Intent intent = PersonActivity.buildIntentFor(activity, contact);

        String transitionName = avatar.getResources().getString(R.string.transition_person_avatar);
        ActivityOptions transitionActivityOptions = ActivityOptions.makeSceneTransitionAnimation(activity, avatar, transitionName);
        activity.startActivity(intent, transitionActivityOptions.toBundle());
        analytics.trackContactDetailsViewed(contact);
    }
}

File: android_mobile/src/main/java/com/alexstyl/specialdates/search/SearchNavigator.kt
package com.alexstyl.specialdates.search

import android.app.Activity
import android.app.ActivityOptions
import android.content.Intent
import android.support.v4.view.ViewCompat
import android.view.View
import com.alexstyl.specialdates.analytics.Analytics
import com.alexstyl.specialdates.analytics.Screen
import com.alexstyl.specialdates.contact.Contact
import com.alexstyl.specialdates.date.Date
import com.alexstyl.specialdates.events.namedays.activity.NamedayActivity
import com.alexstyl.specialdates.permissions.ContactPermissionActivity
import com.alexstyl.specialdates.person.PersonActivity

internal class SearchNavigator(private val analytics: Analytics) {

    fun toContactDetails(contact: Contact, activity: Activity, avatarView: View) {
        val intent = PersonActivity.buildIntentFor(activity, contact)
        val options = ActivityOptions.makeSceneTransitionAnimation(
                activity, avatarView, ViewCompat.getTransitionName(avatarView))
        activity.startActivity(intent, options.toBundle())
        analytics.trackContactDetailsViewed(contact)
    }

    fun toNamedays(date: Date, activity: Activity) {
        val currentYearDate = Date.on(date.dayOfMonth, date.month, Date.CURRENT_YEAR)
        val intent = NamedayActivity.getStartIntent(activity, currentYearDate)
        activity.startActivity(intent)
    }

    fun toContactPermission(activity: Activity) {
        val intent = Intent(activity, ContactPermissionActivity::class.java)
        analytics.trackScreen(Screen.CONTACT_PERMISSION_REQUESTED)
        activity.startActivity(intent)
    }
}
